SilKit/source/services/logging/MessageTracing.hpp

```diff
@@ -24,26 +24,80 @@ WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE. */
 #include "ILoggerInternal.hpp"
 #include "IServiceEndpoint.hpp"
 #include "ServiceDescriptor.hpp"
+#include "traits/SilKitMsgTraits.hpp"
+
 
 namespace SilKit {
 namespace Services {
 
+
+template <typename MsgT>
+std::chrono::nanoseconds GetTimestamp(MsgT& msg,
+                                      std::enable_if_t<Core::HasTimestamp<MsgT>::value, bool> = true)
+{
+    return msg.timestamp;
+}
+
+template <typename MsgT>
+std::chrono::nanoseconds GetTimestamp(MsgT& /*msg*/,
+                  std::enable_if_t<!Core::HasTimestamp<MsgT>::value, bool> = false)
+{
+    return std::chrono::nanoseconds::duration::min();
+}
+
+
+
+
 template <class SilKitMessageT>
-void TraceRx(Logging::ILogger* logger, const Core::IServiceEndpoint* addr, const SilKitMessageT& msg,
+void TraceRx(Logging::ILoggerInternal* logger, const Core::IServiceEndpoint* addr, const SilKitMessageT& msg,
+
              const Core::ServiceDescriptor& from)
 {
-    Logging::Trace(logger, "Recv on {} from {}: {}", addr->GetServiceDescriptor(), from.GetParticipantName(), msg);
+    Logging::LoggerMessage lm{logger, Logging::Level::Trace};
+    lm.SetMessage("Recv message");
+
+    for (const auto& pair : addr->GetServiceDescriptor().to_keyValues())
+    {
+        lm.SetKeyValue(pair.first, pair.second);
+    }
+    lm.SetKeyValue(Logging::Keys::from, from.GetParticipantName());
+    lm.SetKeyValue(Logging::Keys::msg, (fmt::format("{}", msg)));
+
+    auto virtualTimeStamp = GetTimestamp(msg);
+    if (virtualTimeStamp != std::chrono::nanoseconds::duration::min())
+    {
+        lm.SetKeyValue(Logging::Keys::virtualTimeNS, (fmt::format("{}", virtualTimeStamp.count())));
+    }
+
+    lm.Dispatch();
 }
 
 template <class SilKitMessageT>
-void TraceTx(Logging::ILogger* logger, const Core::IServiceEndpoint* addr, const SilKitMessageT& msg)
+void TraceTx(Logging::ILoggerInternal* logger, const Core::IServiceEndpoint* addr, const SilKitMessageT& msg)
 {
-    Logging::Trace(logger, "Send from {}: {}", addr->GetServiceDescriptor(), msg);
+    Logging::LoggerMessage lm{logger, Logging::Level::Trace};
```

Comment:
this will be a performance hit.
The previous call to `Logging::Trace` would not evaluate its arguments, in case the log level is below tracing.
The current function body will be evaluated in any case, right? the TraceTx and TraceRx will be potentially be evaluated for each SilKitMessageT, right?

Reply:
fixed by adding log level checks in all methods. in case this doesn't get optimized away, we might add a top level log level check here.